---
title: "UT23_snRNAseq"
format: html
editor: visual
---

```{r}
library(tidyverse)
library(Seurat)
library(patchwork)
library(hdf5r) #for reading in h5 files 
library(sctransform) #Dhivya recommends using sctransorm for data transformation 
```

```{r}
# loading in the h5 file that was generated by the 10x genomics cloud aggregate function
ut23_data <- Read10X_h5("/stor/work/Fonken/UT23_snRNAseq/filtered_feature_bc_matrix.h5")

# creating seurat object
ut23_object <- CreateSeuratObject(counts = ut23_data, project = "ut23", min.cells = 3, min.features = 300)

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
# This is identifying mitochondrial genes 
ut23_object[["percent.mt"]] <- PercentageFeatureSet(ut23_object, pattern = "^Mt-")

# Visualize QC metrics as a violin plot
VlnPlot(ut23_object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# nFeature_RNA is the number of unique genes (features) in each cell 
# nCount_RNA is the number of RNA molecules detected in each cell
# percent.mt is the percentage of reads that map to mitochondrial genes out of the total reads in each cell 

# Some more plots comparing feature-feature relationships 

plot1 <- FeatureScatter(ut23_object, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(ut23_object, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

ut23_filtered <- subset(ut23_object, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)


```
